Hypergol page (#290)
Nuclear-Squid/ergol: www/content/alternatives/hypergol/index.md

+++
title = "Hypergol"
url = "/hypergol"

[params]
footer = "outillé par [x-keyboard](https://onedeadkey.github.io/x-keyboard)"
+++

Une disposition et configuration de clavier aussi efficace qu’expérimentale !

> ![](hypergol.svg)

Reviewer: Je ne vois pas le É sur l’image.

:::{.highlight style="max-width: 32em;"}
- [Principe de la disposition]
- [La géométrie « Hummingbird »]
- [Le [E]{.kbd} et « touche magique » sous un pouce]
- [Les auto-fills]
- [Les combos étendus]
- [« C’est où qu’on signe ? »]
:::

Principe de la disposition
--------------------------------------------------------------------------------

Hypergol est un projet de disposition de clavier qui cherche à incorporer les
fonctionalités des claviers mécaniques programmables au sein de la disposition.

> À l’heure actuelle, aucune de ses fonctionnalités n’est inderdite : si c’est
> *techniquement faisable* et que ça fait progresser l’ergonomie ou
> l’optimisation de la disposition, alors on y a droit.

Reviewer 2: ```suggestion
À l’heure actuelle, on ne s’interdit aucune de ces fonctionnalités : si c’est
*techniquement faisable* et que ça fait progresser l’ergonomie ou
l’optimisation de la disposition, alors on y a droit.
```

> La disposition est implémentée gràce a un pilote Kanata / QMK prévu pour être

Reviewer 2: ```suggestion
La disposition est implémentée gràce a un pilote Kanata ou QMK prévu pour être
```

utilisé sur un ordinateur en Ergo‑L.

Pour accélérer le développement de la dispo, nous utilisons un [brute-forceur]
custom, qui ne cherche pas à concevoir la meilleur disposition possible sur la
base d’une note globale hazardeuse, mais liste toutes les dispositions
possibles qui addhèrent a des critères simples (pas moins de 12% de charge sur

> un anulaire, pas plus de 0.2% de SFU sur un auriculaire…). On reste sur la

Author: Alors là, j’ai réussi à écrire « annulaire » au lieu de « index», chuis vraiment con

:::{style="display: flex; justify-content: center"}
![Un exemple de clavier « hummingbird »](hummingbird.jpg){width="25em"}
:::

<!-- Bien que le clavier présenté utilise 4 touches de pouces, on cherche aussi a -->
<!-- prévoir une disposition utilisable sur un clavier avec seulement 3 touches de -->
<!-- pouces (comme avec Arsenik). -->

Retirer 4 touches sur le clavier nous force donc à trouver des nouvelles places
pour les 4 lettres les moins fréquentes en français et anglais (`z`, `x`, `k`
et `j`), et pour ça, nous avons décidé de les placer sur la touche typo. On
considère que l’effort nécessaire pour taper ces lettres en deux touches est
négligeable par rapport à l’emplacement où ces lettres peuvent être placées, et
les ciseaux qui peuvent exister.

Il reste encore à savoir comment gérer proprement les touches manquantes en
[AltGr]{.kbd}, on travaille sur la question.

Le [E]{.kbd} et « touche magique » sous un pouce
--------------------------------------------------------------------------------

Puisque nous avons prévu la disposition pour des claviers programables, alors
on peut se permettre de mettre des lettres sous les pouces. Le `e` étant la

> lettre la plus fréquente en français et anglais, c’est donc celle-ci que nous
> avons choisie. Nous avons cependant remarqué que d’avoir le `e` sous le pouce

Reviewer 2: ```suggestion
lettre la plus fréquente en français et anglais, c’est celle-ci qui a été
retenue. Nous avons cependant remarqué qu’avoir le `e` sous le pouce
```

transforme beaucoup de roulements en alternances (avec un pouce), ce qui baisse
drastiquement le taux de roulements de la disposition. On compte faire des

> recherches pour trouver la lettre optimale a cette position.

Reviewer 2: ```suggestion
recherches pour trouver la lettre optimale à cette position.
```

> Passer une lettre sous un pouce permet donc de réqupérer une place dans le pavé
> de 3x10 pour le `é`, et puisque nous n’avons plus besoin mettre le tiret sous
> le `e` (comme en Ergo‑L), on l’a donc échangé avec l’apostrophe typographique
> sur la touche typo, ce qui permet de drastiquement réduire l’usage de cette
> touche typo (~4% -> ~1%).

Reviewer 2: ```suggestion
Passer une lettre sous un pouce permet de récupérer une place dans le pavé
de 3x10 pour le `é` ; et puisque nous n’avons plus besoin de mettre le tiret sous
le `e` comme en Ergo‑L, on l’a donc échangé avec l’apostrophe typographique
sur la touche typo. Cela permet de réduire drastiquement l’usage de cette
touche typo, qui passe de ~4% à ~1% en français.
```

C’est un début mais je crois que c’est **LE** paragraphe à reformuler. J’imagine que l’idée est de dire que la combinaison Typo+Espace fait désormais un tiret en non plus une apostrophe (qui est désormais disponible en direct), mais ça n’est pas clair en l’état.

La « touche magique » (inspirée de [Magic Sturdy]) est une touche dont le

> comportement dépend de la touche précédente (un peu comme l’inverse d’une
> touche morte). Elle se comporte comme une touche « [alt repeat] » de QMK : elle

Reviewer 2: ```suggestion
comportement dépend de la touche précédente — un peu comme l’inverse d’une
touche morte. Elle se comporte comme une touche « [alt repeat] » de QMK : elle
```

répète les symboles fréquemment doublés et fait un autre symbole sur le même

> doigt pour les autres. L’objectif de cette touche est de totalement (ou
> presque) éliminer les SKB et SFB.

Reviewer 2: ```suggestion
doigt pour les autres. L’objectif de cette touche est d’éliminer totalement
(ou presque) les SKB et SFB.
```

> Par exemple, `ui` est un SFB qui a 0.7% de fréquence en français, donc après un

Reviewer 2: ```suggestion
Par exemple, `ui` est un SFB qui a 0,7 % de fréquence en français, donc après un
```

`u`, la touche magique insère un `i`. De nombreux SFB sont *intentionnellement*
inséré dans la disposition pour qu’ils soient corrigés par cette touche
magique. Cela permet d’avoir à la fois un score de SFU très bas (estimé à 0.3%

> en fraçais et 0.4% en anglais) tout en se laissant assez de marge de manœuvre

Reviewer 2: ```suggestion
en français et 0.4% en anglais) tout en se laissant assez de marge de manœuvre
```

> magique. Cela permet d’avoir à la fois un score de SFU très bas (estimé à 0.3%
> en français et 0.4% en anglais) tout en se laissant assez de marge de manœuvre

Reviewer 2: ```suggestion
magique. Cela permet d’obtenir un score de SFU très bas (estimé à 0,3 %
en français et 0,4 % en anglais) tout en se laissant assez de marge de manœuvre
```

Le « à la fois… » est redondant avec le « tout en se laissant… » qui suit.

pour limiter les ciseaux, LSB et mauvaises redirections.

La touche magique agit comme une touche repeat par défaut, sauf pour les
lettres suivantes :

|       |     |     |     |               |     |     |     |      |     |     |     |     |     |
| ----- | --- | --- | --- | ---           | --- | --- | --- | ---  | --- | --- | --- | --- | --- |
| Prev  | `A` | `C` | `D` | `E`           | `G` | `H` | `I` | `Q`  | `U` | `V` | `É` | `’` | `Y` |
| Magic | `O` | `★` | `Y` | [Space]{.kbd} | `T` | `.` | `,` | `U’` | `I` | `R` | `A` | `T` | `D` |

### Les configurations de pouces recommandés

> Bien que la dispo soit avec une configuration de pouces précise, vous êtes
> libre d’adapter cette configuration comme bon vous semble. Il reste cependant
> quelques pièges à éviter.

Reviewer 2: ```suggestion
Bien que la dispo soit proposée avec une configuration de pouces précise, vous
êtes libre d’adapter cette configuration comme bon vous semble. Il y a cependant
quelques pièges à éviter.
```

Nous recommandont d’utiliser la configuration d’[Arsenik] ou [Selenium]
(suivant le nombre de touches de pouce) pour les actions en hold. En tap, nous
recommandont de placer la touche magique sur le pouce opposé du `e` (car il est
fréquemment doublé en anglais) et sur le même pouce que l’espace (pour éviter

> les SFB `e` -> `espace`). Idéalement, l’espace devrait être accessible par les

Reviewer: Dans le tableau au dessus, il est dit que `E magic` fait  fait `E Space`, ce qui est incohérent avec ce paragraphe.

Malheureusement, avoir une lettre fréquemment doublée sous un pouce peut rendre
la configuration des pouces compliquée : devoir mettre la touche magique sous
le même pouce que l’espace nous force à avoir des configurations complexes et
peu fiables pour totalement éliminer les SKB et SFK de pouce (qu’on a remarqué
être particulièrement inconfortables) et profiter pleinement des touches de
pouces disponibles (par exemple, temporairement transformer le one-time-shift
en espace après une pression de la touche magique).

Avoir espace et one-time-shift d’un côté et une lettre (non doublée) et magic
de l’autre permettrait d’éliminer ces soucis de configuration. Nous travaillons
sur cette question, mais la solution peut prendre encore du temps.

Les auto-fills
--------------------------------------------------------------------------------

Les auto-fills sont une fonctionnalité exclusive à Hypergol, leur objectif est
d’éliminer les mauvaises redirections et économiser quelques touches.

### Un exemple concret

Une façon de comprendre le concept de cette fonctionnalité est de regarder le
fonctionnement de la touche `Qu`.

> En français comme en anglais, le `Q` n’est pratiquement *que* suivi par un `u`,
> donc une façon d’économiser des gestes est d’utiliser une touche qui écrit
> directement `qu`. Cependant, cela peut causer des problèmes, si on veut écrire
> `Qatar`, `cinq` ou commencer une phrase par un `q` (puisque shift + `qu` donne
> `QU` au lieu de `Qu`). Les auto-fills sont une façon beaucoup plus fiable
> d’implémenter une touche `Qu`.

Reviewer 2: ```suggestion
En français comme en anglais, le `Q` étant quasiment *toujours* suivi par un `u`,
une façon d’économiser des gestes serait donc d’utiliser une touche qui écrit
directement `qu`. Mais cela peut causer des problèmes, notamment pour écrire
`Qatar`, `cinq` ou commencer une phrase par un `q` (puisque shift + `qu` donne
`QU` et non `Qu`). Pour éviter ces écueils, on a recours à des auto-fills
plutôt qu’à une touche `Qu`.
```

> Une pression sur la touche `q` va immédiatement écrire un `q` et va surveiller

Reviewer 2: ```suggestion
Une pression sur la touche `q` va immédiatement écrire un `q` et surveiller
```

la touche suivante. Si la touche suivante est fait partie de `aeio’é`, alors on
rajoute un `u` juste avant d’écrire la lettre voulue (et si la touche ne fait

> pas partie de la liste, alors on sontinue comme si de rien était). On peut donc

Reviewer: ```suggestion
pas partie de la liste, alors on continue comme si de rien était). On peut donc
```

> donc facilement écrire `Que` ou `cinq` sans problèmes. Ça reste plus compliqué

Reviewer 2: ```suggestion
donc facilement écrire `Que` ou `cinq` sans problème. Ça reste plus compliqué
```

d’écrire `Qatar` mais pour l’instant, on considère que faire typo -> `q` insère
un `q` sans l’auto-fill.

### Une définition plus formelle

Si un roulement sur les lettres quelconques `a` et `c` a une fréquence
d’apparition négligeable, alors on peut s’en servir pour taper un auto-fill
`abc`. Cela permet à la fois d’économiser une touche mais aussi de limiter les
mauvaises redirections.

> Par exemple, dans la version actuelle d’Hypergol, on a une mauvaise redirection
> sur `you` (en anglais). Cependant, le bigramme `yu` n’existe pratiquement pas,
> donc on peut utiliser un auto-fill pour écrire `you` en tapant `yu`, ce qui
> transforme une mauvaise redirection en un roulement intérieur.

Reviewer 2: ```suggestion
Par exemple, dans la version actuelle d’Hypergol, on a une mauvaise redirection
sur `you` (en anglais). Mais puisque le bigramme `yu` n’existe pratiquement pas,
on peut utiliser un auto-fill pour écrire `you` en tapant `yu`, et ainsi
transformer une mauvaise redirection en un roulement intérieur.
```

> Les auto-fills sont nés en cherchant une meilleur implémentation de la touche
> `qu`, et ce n’est que plus tard que nous avons trouvé cet usage, donc Hypergol
> ne profite pas encore pleinement des posibilités que cela ouvre.

Reviewer 2: ```suggestion
Les auto-fills sont nés en cherchant une meilleurs implémentation de la touche
`qu`, et ce n’est que plus tard que nous avons trouvé cet usage pour `you`.
Hypergol ne profite donc pas encore pleinement des possibilités que cela ouvre.
```

Les « combos étendus »
--------------------------------------------------------------------------------

Les « combos étendus » sont une autre fonctionnalité exclusive à Hypergol, et
c’est probablement la partie la plus complexe de la disposition. Cette
fonctionnalité permet de transformer la dispo en un hybride entre un clavier
standard et un clavier de sténotypie.

> L’idée est qu’on peut taper un combo (appuyer sur deux touches spécifiques en
> même temps) pour écrire un mot ou morceau de mot. Comme pour les auto-fills, la

Reviewer 2: ```suggestion
L’idée est qu’on peut taper un combo (deux touches spécifiques actionnées en
même temps) pour écrire un mot ou morceau de mot. Comme pour les auto-fills, la
```

touche suivante va être surveillée ; mais cette fois, au lieu de rajouter une
seule lettre, on rentre dans une machine a états qui va écrire le reste du mot
et potentiellement continuer la séquence.

> Par exemple, appuyer sur [S]{.kbd} et [F]{.kbd} (de Qwerty) en même temps, va
> écrire `the`, et donne accès a ces extensions suivantes :

Reviewer 2: ```suggestion
Par exemple, appuyer sur [S]{.kbd} et [F]{.kbd} (de Qwerty) en même temps va
écrire `the`, et donne accès aux extensions suivantes :
```

|         |        |         |         |           |           |        |
|   ---   | ---    | ---     | ---     | ---       | ---       | ---    |
| lettre  | `F`    | `N`     | `S`     | `R`       | `L`       | `M`    |
| mot     | `they` | `there` | `their` | `they’re` | `they’ll` | `them` |

(Comme pour les auto-fills, écrire n’importe quelle autre lettre va sortir de

> la machine a états et écrire la touche comme si de rien n’était)

Reviewer: Du coup, la définition de `M` qui fait `them` est inutile, non ?

À l’heure actuelle, c’est le seul combo étendu de la dispo, il sert surtout de

> « proof of concept » et bien que l’expérience soit satisfaisante, nous en avons

Reviewer 2: ```suggestion
« proof of concept » et bien que l’expérience soit satisfaisante, nous n’en avons
```

pas encore rajouté d’autres : il reste des parties plus critiques à régler
avant ça.

« C’est où qu’on signe ? »
--------------------------------------------------------------------------------

Certains d’entre vous doivent probablement beaucoup aimer le concept, et

> voudront utiliser la disposition le plus tôt possible. Cependant, comme cela a
> été mentionné quelques fois déjà, c’est encore un travail en cours : Il y a des

Reviewer 2: ```suggestion
voudront utiliser la disposition le plus tôt possible. Mais attention, c’est en plein
développement et ça n’est pas encore prévu pour une utilisation quotidienne : il y a des
```

bugs dans les pilotes, notre analyseur ne sait pas gérer ce type de

> dispositions, il manque des features dans beaucoup d’applis / frameworks pour

Reviewer 2: ```suggestion
dispositions, il manque des fonctionnalités dans beaucoup d’applis/frameworks pour
```

claviers programmables qu’on doit implémenter nous mêmes, il faut travailler

> sur le [brute-forceur custom] pour qu’il puisse repérer les mauvaises

Reviewer 2: ```suggestion
sur le [brute-forceur maison] pour qu’il puisse repérer les mauvaises
```

redirections (et proposer des auto-fills), la disposition peut encore

> radicalement bouger et nous ne sommes pas a l’abris de découvrir une nouvelle

Reviewer 2: ```suggestion
radicalement bouger et nous ne sommes pas a l’abri de découvrir une nouvelle
```

> Certains d’entre vous doivent probablement beaucoup aimer le concept, et
> voudront utiliser la disposition le plus tôt possible. Mais attention, c’est en
> plein développement et ça n’est pas encore prévu pour une utilisation
> quotidienne : il y a des bugs dans les pilotes, notre analyseur ne sait pas
> gérer ce type de dispositions, il manque des fonctionnalités dans beaucoup
> d’applis/frameworks pour claviers programmables qu’on doit implémenter nous
> mêmes, il faut travailler sur le [brute-forceur maison] pour qu’il puisse
> repérer les mauvaises redirections (et proposer des auto-fills), la disposition
> peut encore radicalement bouger et nous ne sommes pas a l’abri de découvrir une
> nouvelle fonctionnalité, ou remettre en cause une fonctionnalité existante.

Reviewer 2: ```suggestion
Certains d’entre vous doivent probablement beaucoup aimer le concept, et
voudront utiliser la disposition le plus tôt possible. Mais attention, c’est en
plein développement et ça n’est pas encore prévu pour une utilisation
quotidienne : il y a des bugs dans les pilotes, notre analyseur ne sait pas
gérer ce type de dispositions, il manque des fonctionnalités dans beaucoup
d’applis/frameworks pour claviers programmables qu’on doit implémenter
nous-mêmes, il faut travailler sur le [brute-forceur maison] pour qu’il puisse
repérer les mauvaises redirections (et proposer des auto-fills), la disposition
peut encore radicalement bouger — et nous ne sommes à l’abri ni de découvrir
une nouvelle fonctionnalité, ni de remettre en cause une fonctionnalité existante.
```
